### PR TITLE
[wasm] Support network status information

### LINKED ISSUE #38988: [wasm] Implement NetworkInterface.GetIsNetworkAvailable and NetworkChange.NetworkAvailabilityChanged
Since we don't have access to the network cards on browser/WASM, the entire System.Net.NetworkInformation assembly is going to be PNSE excepting tho methods:
- NetworkInterface.GetIsNetworkAvailable
- NetworkChange.NetworkAvailabilityChanged

These two methods could probably be implemented on top of https://developer.mozilla.org/en-US/docs/Web/API/NavigatorOnLine/onLine

### PR DESCRIPTION
**Pull request by maraf**
Support these network status related API on browser

- `NetworkInterface.GetIsNetworkAvailable`
- `NetworkChange.NetworkAvailabilityChanged`

TODO
- [x] Resolve how it should behave when the browser API is missing
- [ ] ~~Try to control a devtools network status in unit test~~
- [ ] Try to use playwright and WBT for unit test
- [ ] Try blazor sdk static assets test

Closes https://github.com/dotnet/runtime/issues/38988

### CODE REVIEW

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Browser.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        private static event NetworkAvailabilityChangedEventHandler? s_networkAvailabilityChanged;
+
+        [UnsupportedOSPlatform("illumos")]
+        [UnsupportedOSPlatform("solaris")]
+        public static event NetworkAvailabilityChangedEventHandler? NetworkAvailabilityChanged
+        {
+            add
+            {
+                if (s_networkAvailabilityChanged == null)
```
Comment (pavelsavara) [inline src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Browser.cs:24]: this needs to happen on main browser thread I guess. Shall we add guard for it ?

**Comment by maraf** [comment]
The browser API is not supported on V8 and NodeJS. I'm not sure how to deal with this in general, throw or let it fail in JS?

cc @pavelsavara @lewing @kg

**Comment by kg** [comment]
> The browser API is not supported on V8 and NodeJS. I'm not sure how to deal with this in general, throw or let it fail in JS?
> 
> cc @pavelsavara @lewing @kg

I would throw. Then tests can cleanly check for it

**Comment by maraf** [comment]
> I would throw. Then tests can cleanly check for it

Throw when the API is not available at runtime or throw "statically" when on V8/nodejs? Do we have any general guidance for optional browser APIs?

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Browser.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+using System.ComponentModel;
+using System.Collections.Generic;
+using System.Net.Sockets;
+using System.Runtime.Versioning;
+using System.Threading;
+
+namespace System.Net.NetworkInformation
+{
+    public partial class NetworkChange
```
Comment (radical) [inline src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Browser.cs:14]: Why not do this like other platform specific partial classes like `NetworkAddressChange.Unix.cs`?

File: src/libraries/System.Net.NetworkInformation/tests/FunctionalTests/AddressParsingTests.cs [hunk 3]
```diff
@@ -7,6 +7,7 @@
 
 namespace System.Net.NetworkInformation.Tests
 {
+    [SkipOnPlatform(TestPlatforms.Browser, "Not supported on Browser")]
```
Comment (radical) [inline src/libraries/System.Net.NetworkInformation/tests/FunctionalTests/AddressParsingTests.cs:10]: Does parsing addresses need something platform specific that isn't available on the browser?

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/IPGlobalPropertiesPal.Browser.cs [hunk 4]
```diff
@@ -0,0 +1,10 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Net.NetworkInformation
+{
+    internal static class IPGlobalPropertiesPal
+    {
+        public static IPGlobalProperties GetIPGlobalProperties() => throw new PlatformNotSupportedException();
```
Comment (radical) [inline src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/IPGlobalPropertiesPal.Browser.cs:8]: unused?

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkInterfacePal.Browser.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+        [JSImport("INTERNAL.network_wasm_add_change_listener")]
+        public static partial void AddChangeListener([JSMarshalAs<JSType.Function<JSType.Boolean>>] Action<bool> handler);
+    }
+
+    internal static class NetworkInterfacePal
+    {
+        public static NetworkInterface[] GetAllNetworkInterfaces() => throw new PlatformNotSupportedException();
+        public static int IPv6LoopbackInterfaceIndex => throw new PlatformNotSupportedException();
+        public static int LoopbackInterfaceIndex => throw new PlatformNotSupportedException();
```
Comment (radical) [inline src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkInterfacePal.Browser.cs:21]: Why are these needed? Also, they don't seem to be used.

File: src/mono/wasm/runtime/exports.ts [hunk 6]
```diff
@@ -485,6 +486,10 @@ const INTERNAL: any = {
     http_wasm_get_response_bytes,
     http_wasm_get_response_length,
     http_wasm_get_streamed_response_bytes,
+
+    // NetworkInformation
+    network_wasm_online,
```
Comment (radical) [inline src/mono/wasm/runtime/exports.ts:491]: nit: `network_wasm_is_online`

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Browser.cs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+                s_networkAvailabilityChanged += value;
+            }
+            remove
+            {
+                s_networkAvailabilityChanged -= value;
+            }
+        }
+
+        private static void OnNetworkChanged(bool isOnline) => s_networkAvailabilityChanged?.Invoke(null, new NetworkAvailabilityEventArgs(isOnline));
```
Comment (radical) [inline src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Browser.cs:39]: nit: 
```suggestion
        private static void OnNetworkAvailabilityChanged(bool isOnline) => s_networkAvailabilityChanged?.Invoke(null, new NetworkAvailabilityEventArgs(isOnline));
```

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs [hunk 8]
```diff
@@ -39,6 +39,7 @@ public partial class NetworkChange
         private static readonly AutoResetEvent s_runLoopStartedEvent = new AutoResetEvent(false);
         private static readonly AutoResetEvent s_runLoopEndedEvent = new AutoResetEvent(false);
 
+        [UnsupportedOSPlatform("browser")]
```
Comment (radical) [inline src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs:42]: IIUC, this shouldn't be needed in platform specific implementation files, since they are included in the build only for the relevant platforms. Same for others.

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Browser.cs [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+        [UnsupportedOSPlatform("browser")]
+        [UnsupportedOSPlatform("illumos")]
+        [UnsupportedOSPlatform("solaris")]
+        public static event NetworkAddressChangedEventHandler? NetworkAddressChanged
+        {
+            add => throw new System.PlatformNotSupportedException(System.SR.SystemNetNetworkInformation_PlatformNotSupported);
+            remove => throw new System.PlatformNotSupportedException(System.SR.SystemNetNetworkInformation_PlatformNotSupported);
```
Comment (radical) [inline src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Browser.cs:47]: The corresponding tests should be run for browser, and check for PNSE.

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkInterfacePal.Browser.cs (hunk 5 above)
Comment (maraf) [reply src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkInterfacePal.Browser.cs:21]: They are used here https://github.com/dotnet/runtime/blob/2baf83fd2a162f75850885733bf63c2fcbd97c46/src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkInterface.cs#L19

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Browser.cs [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        private static bool s_isBrowserNetworkChangeListenerAttached;
+        private static event NetworkAvailabilityChangedEventHandler? s_networkAvailabilityChanged;
+
+        [UnsupportedOSPlatform("illumos")]
+        [UnsupportedOSPlatform("solaris")]
+        public static event NetworkAvailabilityChangedEventHandler? NetworkAvailabilityChanged
+        {
+            add
+            {
+                if (!s_isBrowserNetworkChangeListenerAttached)
```
[the remark this answers is not in the document]
Comment (maraf) [reply src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Browser.cs:25]: This should happen on the main thread, but don't have a way to check for it yet

cc @lambdageek

File: src/libraries/System.Net.NetworkInformation/tests/FunctionalTests/NetworkAvailabilityChangedTests.cs [hunk 11]
```diff
@@ -7,6 +7,7 @@ namespace System.Net.NetworkInformation.Tests
 {
     // Partial class used for both NetworkAddressChanged and NetworkAvailabilityChanged
     // so that the tests for each don't run concurrently
+    [SkipOnPlatform(TestPlatforms.Browser, "Not supported on Browser")]
```
[the remark this answers is not in the document]
Comment (maraf) [reply src/libraries/System.Net.NetworkInformation/tests/FunctionalTests/NetworkAvailabilityChangedTests.cs:10]: First of them yes, the rest are using `NetworkAddressChanged`. Anyway, I would like try to write tests that toggle network state in devtools.

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Browser.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (pavelsavara) [reply src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Browser.cs:25]: It would be good to add some guard which will fail on other threads, until we know how to dispatch it onto main thread. 
Same would be good for other things like WS or HTTP.

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs (hunk 8 above)
Comment (radical) [reply src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs:42]: Make sure to include the main .cs file also. What was the output from apicompat?

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Browser.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (maraf) [reply src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Browser.cs]: I agree, but according to discussion with Aleksey, we don't have anything yet available in C# and probably this PR is not a place to add it.

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs (hunk 8 above)
Comment (maraf) [reply src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs:42]: If I just drop this one and try to compile with 
```
.\dotnet.cmd build /p:TargetOS=Browser /p:TargetArchitecture=wasm /p:Configuration=Debug .\src\libraries\System.Net.NetworkInformation\src\
```
it fails with
```
CannotChangeAttribute : Attribute 'System.Runtime.Versioning.UnsupportedOSPlatformAttribute' on 'System.Net.NetworkInformation.NetworkChange
.NetworkAddressChanged' changed from '[UnsupportedOSPlatformAttribute("browser")]' in the contract to '[UnsupportedOSPlatformAttribute("illumos")]' in the implementation. [....\runtime\src\libraries\System.Net.NetworkInformation\src\System.Net.NetworkInformation
.csproj]
```

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Browser.cs (hunk 2 above)
Comment (maraf) [reply src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Browser.cs:14]: I don't understand what you mean. Can you please give me a clue?

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Browser.cs (hunk 9 above)
Comment (maraf) [reply src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Browser.cs:47]: Just curious, are we writing a test for every API that is not supported on any platform?
Because this is already not supported on `illumos` and `solaris`. Should I run the test also on these platforms?

File: src/mono/wasm/runtime/network-information.ts [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+
+export function network_wasm_is_online(): boolean {
+    return globalThis.navigator.onLine;
+}
+
+let change_listener: ((onLine: boolean) => void) | null = null;
+
+export function network_wasm_set_change_listener(listener: (onLine: boolean) => void): void {
+    if (window) {
```
Comment (pavelsavara) [inline src/mono/wasm/runtime/network-information.ts:11]: `assert(globalThis.window)`
